include/dmlc/io.h

```diff
@@ -205,19 +227,56 @@ class InputSplit {
    * \param part_index the part id of current input
    * \param num_parts total number of splits
    * \param type type of record
-   *   List of possible types: "text", "recordio"
+   *   List of possible types: "text", "recordio", "indexed_recordio"
    *     - "text":
    *         text file, each line is treated as a record
    *         input split will split on '\\n' or '\\r'
    *     - "recordio":
    *         binary recordio file, see recordio.h
+   *     - "indexed_recordio":
+   *         binary recordio file with index, see recordio.h
    * \return a new input split
    * \sa InputSplit::Type
    */
   static InputSplit* Create(const char *uri,
                             unsigned part_index,
                             unsigned num_parts,
                             const char *type);
+  /*!
+   * \brief factory function:
+   *  create input split given a uri for input and index
+   * \param uri the uri of the input, can contain hdfs prefix
+   * \param index_uri the uri of the index, can contain hdfs prefix
+   * \param part_index the part id of current input
+   * \param num_parts total number of splits
+   * \param type type of record
+   *   List of possible types: "text", "recordio", "indexed_recordio"
+   *     - "text":
+   *         text file, each line is treated as a record
+   *         input split will split on '\\n' or '\\r'
+   *     - "recordio":
+   *         binary recordio file, see recordio.h
+   *     - "indexed_recordio":
+   *         binary recordio file with index, see recordio.h
+   * \param shuffle whether to shuffle the output from the InputSplit,
+   *                supported only by "indexed_recordio" type.
+   *                Defaults to "false"
+   * \param seed random seed to use in conjunction with the "shuffle"
+   *             option. Defaults to 0
+   * \param batch_size a hint to InputSplit what is the intended number
+   *                   of examples return per batch. Used only by
+   *                   "indexed_recordio" type
+   * \return a new input split
+   * \sa InputSplit::Type
+   */
+  static InputSplit* Create(const char *uri,
+                            const char *index_uri,
+                            unsigned part_index,
+                            unsigned num_parts,
+                            const char *type,
+                            const bool shuffle = false,
+                            const int seed = 0,
+                            const size_t batch_size = 256);
```

Comment:
This is only a hint (like the HintChunkSize) and only IndexedRecordioInputSplit actually uses that. Readers can give back any number of samples anyway so consumers of the read data need to be prepared to handle that.